doc/types.go

```diff
@@ -0,0 +1,55 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package doc
+
+// ID represents a document's natural identifier.
+type ID []byte
+
+// FieldValueType represents the different types of field values supported.
+type FieldValueType int
+
+const (
+	// UnknownValueType marks un-specified value types.
+	UnknownValueType FieldValueType = iota
+
+	// StringValueType demarks string value types.
+	StringValueType
+)
+
+// Field represents a document field.
+type Field struct {
+	Name      []byte
+	Value     []byte
+	ValueType FieldValueType
+}
+
+// Document represents a document to be indexed.
+type Document interface {
+	// ID is the natural identifier for the document.
+	ID() ID
+
+	// IndexFields contain the list of fields to index by.
```

Comment:
`Fields` is a bit too in-specific, we don't want people to put "stored" fields in here too. 

I was trying to avoid putting tense in the name because it implies something about the document's lifecycle. The same document object is in use prior to ingest (which makes `IndexedFields` weird), and once the doc has been ingested (which makes `IndexableFields` weird). Does that make sense?

Reply:
We should also talk about how we want to handle `StoredFields`

Reply:
Agreed to call it `IndexableFields`

Reply:
going to stick with Bytes() as the only "stored" field for now. Can model stored fields first class down the road if needed